### PR TITLE
UCS/ASYNC: Remove async_max_events limit

### PR DESCRIPTION
**Pull request by huaxiangfan**
## What
Remove async_max_events limit

## Why ?
To support unexpected high number of concurrent TCP connections for example

## How ?
By using ucs_queue instead of array inside mpmc queue

### CODE REVIEW

File: src/ucs/async/async.c [hunk 1]
```diff
@@ -436,12 +436,8 @@ ucs_async_alloc_handler(int min_id, int max_id, ucs_async_mode_t mode,
         goto err;
     }
 
-    /* Limit amount of handlers per context */
     if (async != NULL) {
-        if (ucs_atomic_fadd32(&async->num_handlers, 1) >= ucs_global_opts.async_max_events) {
-            status = UCS_ERR_EXCEEDS_LIMIT;
-            goto err_dec_num_handlers;
-        }
+        ucs_atomic_fadd32(&async->num_handlers, 1);
```
Comment (yosefe) [inline src/ucs/async/async.c:440]: let's remove num_handlers altogether

File: src/ucs/config/global_opts.c [hunk 2]
```diff
@@ -152,10 +151,6 @@ static ucs_config_field_t ucs_global_opts_table[] = {
    "Enable memory type (cuda/rocm) cache",
    ucs_offsetof(ucs_global_opts_t, enable_memtype_cache), UCS_CONFIG_TYPE_TERNARY},
 
- {"ASYNC_MAX_EVENTS", "1024", /* TODO remove this; resize mpmc */
```
Comment (yosefe) [inline src/ucs/config/global_opts.c:155]: instead of removing , mark it as UCS_CONFIG_TYPE_DEPRECATED

File: src/ucs/datastruct/mpmc.c [hunk 3]
```diff
@@ -17,82 +17,66 @@
...
 {
-    uint32_t i;
+    ucs_status_t ret;
 
-    mpmc->length   = ucs_roundup_pow2(length);
-    mpmc->shift    = ucs_ilog2(mpmc->length);
-    if (mpmc->shift >= UCS_MPMC_VALID_SHIFT) {
-        return UCS_ERR_INVALID_PARAM;
+    ret = ucs_spinlock_init(&mpmc->lock, 0);
+    if (ret == UCS_OK) {
```
Comment (yosefe) [inline src/ucs/datastruct/mpmc.c:25]:
```c
ucs_queue_head_init(&mpmc->queue);
return ucs_spinlock_init(&mpmc->lock, 0);
```

File: src/ucs/datastruct/mpmc.c [hunk 4]
```diff
@@ -17,82 +17,66 @@
...
 void ucs_mpmc_queue_cleanup(ucs_mpmc_queue_t *mpmc)
 {
-    ucs_free(mpmc->queue);
-}
+    ucs_queue_elem_t *elem;
 
-static inline uint64_t __ucs_mpmc_queue_valid_bit(ucs_mpmc_queue_t *mpmc, uint32_t location)
-{
-    return ((uint64_t)location >> mpmc->shift) & 1;
+    while ((elem = ucs_queue_pull(&mpmc->queue))) {
```
Comment (yosefe) [inline src/ucs/datastruct/mpmc.c:36]:
```c
while (!ucs_queue_is_empty(&mpmc->queue)) {
    elem = ucs_queue_pull_elem_non_empty(&mpmc->queue, ucs_mpmc_elem_t, super);
    ucs_free(elem);
}
```

File: src/ucs/datastruct/mpmc.c [hunk 5]
```diff
@@ -17,82 +17,66 @@
...
+    ucs_mpmc_elem_t *elem;
+
+    elem = ucs_malloc(sizeof(ucs_mpmc_elem_t), "mpmc elem");
+    if (elem == NULL) {
+        return UCS_ERR_NO_MEMORY;
+    }
 
-    ucs_assert((value >> UCS_MPMC_VALID_SHIFT) == 0);
+    elem->value      = value;
+    elem->super.next = NULL;
```
Comment (yosefe) [inline src/ucs/datastruct/mpmc.c:51]: not needed; the linked list is handled by ucs_queue functions

File: src/ucs/datastruct/mpmc.c [hunk 6]
```diff
@@ -17,82 +17,66 @@
...
 {
-    uint32_t location;
-    uint64_t value;
+    ucs_queue_elem_t *queue_elem;
+    ucs_mpmc_elem_t *mpmc_elem;
 
-    location = mpmc->consumer;
-    if (location == mpmc->producer) {
-        /* Producer not started yet */
+    if (!ucs_spin_try_lock(&mpmc->lock)) {
```
Comment (yosefe) [inline src/ucs/datastruct/mpmc.c:66]:
```c
if (ucs_queue_is_empty(&mpmc->queue)) {
    return UCS_ERR_NO_PROGRESS;
}

ucs_spin_lock()
if (!ucs_queue_is_empty(&mpmc->queue)) {
    elem     = ucs_queue_pull_elem_non_empty(..);
    *value_p = elem->value;
    status   = UCS_OK;
    free(elem);
} else {
    status = UCS_ERR_NO_PROGRESS;
}
ucs_spin_unlock(...)
```

File: src/ucs/datastruct/mpmc.h [hunk 7]
```diff
@@ -7,34 +7,36 @@
 #ifndef UCS_MPMC_H
 #define UCS_MPMC_H
 
+#include "queue.h"
+
 #include <ucs/type/status.h>
+#include <ucs/type/spinlock.h>
```
Comment (yosefe) [inline src/ucs/datastruct/mpmc.h:13]: maybe no need to include math.h anymore

File: src/ucs/datastruct/mpmc.h [hunk 8]
```diff
@@ -7,34 +7,36 @@
...
-    uint32_t           length;      /* Array size. Rounded to power of 2. */
-    int                shift;
-    volatile uint32_t  producer;    /* Producer index */
-    volatile uint32_t  consumer;    /* Consumer index */
-    uint64_t           *queue;      /* Array of data */
+    ucs_spinlock_t     lock;        /* Protects 'queue' */
+    ucs_queue_head_t   queue;       /* Queue of data */
 } ucs_mpmc_queue_t;
 
+/**
```
Comment (yosefe) [inline src/ucs/datastruct/mpmc.h:25]: 2 space lines in H file

File: src/ucs/async/async.c (hunk 1 above)
Comment (huaxiangfan) [reply src/ucs/async/async.c]: done

File: src/ucs/config/global_opts.c (hunk 2 above)
Comment (huaxiangfan) [reply src/ucs/config/global_opts.c:155]: done

File: src/ucs/datastruct/mpmc.c (hunk 3 above)
Comment (huaxiangfan) [reply src/ucs/datastruct/mpmc.c]: done

File: src/ucs/datastruct/mpmc.c (hunk 4 above)
Comment (huaxiangfan) [reply src/ucs/datastruct/mpmc.c]: done

File: src/ucs/datastruct/mpmc.c (hunk 5 above)
Comment (huaxiangfan) [reply src/ucs/datastruct/mpmc.c]: done

File: src/ucs/datastruct/mpmc.c (hunk 6 above)
Comment (huaxiangfan) [reply src/ucs/datastruct/mpmc.c]: done

File: src/ucs/datastruct/mpmc.h (hunk 7 above)
Comment (huaxiangfan) [reply src/ucs/datastruct/mpmc.h:13]: done

File: src/ucs/datastruct/mpmc.h (hunk 8 above)
Comment (huaxiangfan) [reply src/ucs/datastruct/mpmc.h:25]: done

File: src/ucs/async/async.c [hunk 9]
```diff
@@ -461,16 +458,16 @@ ucs_async_alloc_handler(int min_id, int max_id, ucs_async_mode_t mode,
         goto err_free;
     }
 
+    if (async != NULL) {
+        ucs_atomic_fadd32(&async->num_handlers, 1);
```
Comment (yosefe) [inline src/ucs/async/async.c:462]: can we delete the field async->num_handlers?

File: src/ucs/async/async.c [hunk 10]
```diff
@@ -14,6 +14,7 @@
 #include <ucs/debug/debug_int.h>
 #include <ucs/datastruct/khash.h>
 #include <ucs/sys/stubs.h>
+#include <ucs/sys/math.h>
```
Comment (yosefe) [inline src/ucs/async/async.c:17]: why do we need math.h here?

File: src/ucs/datastruct/mpmc.c [hunk 11]
```diff
@@ -17,82 +17,62 @@
...
-    uint32_t location;
+    ucs_mpmc_elem_t *elem;
+
+    elem = ucs_malloc(sizeof(ucs_mpmc_elem_t), "mpmc elem");
+    if (elem == NULL) {
+        return UCS_ERR_NO_MEMORY;
+    }
 
-    ucs_assert((value >> UCS_MPMC_VALID_SHIFT) == 0);
+    elem->value      = value;
```
Comment (yosefe) [inline src/ucs/datastruct/mpmc.c:46]: remove extra spaces

File: src/ucs/async/async.c (hunk 9 above)
Comment (huaxiangfan) [reply src/ucs/async/async.c:462]: does it mean removing this if block https://github.com/openucx/ucx/blob/313ba9f7fded9688096660245ae1c574ab6bdd73/src/ucs/async/async.c#L395?

File: src/ucs/async/async.c (hunk 10 above)
Comment (huaxiangfan) [reply src/ucs/async/async.c:17]: ucs_max needs it. since we remove math.h off of mpmc.h, we need to add math.h here.

Comment (yosefe) [reply src/ucs/async/async.c:17]: ok

File: src/ucs/async/async.c (hunk 9 above)
Comment (yosefe) [reply src/ucs/async/async.c:462]: yes

File: src/ucs/datastruct/mpmc.c (hunk 11 above)
Comment (huaxiangfan) [reply src/ucs/datastruct/mpmc.c]: done

File: src/ucs/async/async.c (hunk 9 above)
Comment (huaxiangfan) [reply src/ucs/async/async.c]: done

File: src/ucs/async/async.c [hunk 12]
```diff
@@ -389,22 +388,8 @@ ucs_status_t ucs_async_context_create(ucs_async_mode_t mode,
...
-    if (async->num_handlers > 0) {
-        pthread_rwlock_rdlock(&ucs_async_global_context.handlers_lock);
-        kh_foreach_value(&ucs_async_global_context.handlers, handler, {
-            if (async == handler->async) {
-                ucs_warn("async %p handler "UCS_ASYNC_HANDLER_FMT" not released",
-                         async, UCS_ASYNC_HANDLER_ARG(handler));
-            }
-        });
-        ucs_warn("releasing async context with %d handlers", async->num_handlers);
-        pthread_rwlock_unlock(&ucs_async_global_context.handlers_lock);
```
Comment (yosefe) [inline src/ucs/async/async.c:404]: this part should NOT be removed. just remote the if condition around it.

**Review by yosefe — APPROVED** [review]

**Comment by yosefe** [comment]
@huaxiangfan pls squash

**Comment by huaxiangfan** [comment]
> @huaxiangfan pls squash

done

**Pull request merged**